brayns/io/loaders/protein/ProteinLoader.cpp

```diff
@@ -19,440 +20,223 @@
-                    value = "";
-                    break;
-                case 21:
-                    atom.chainId = (int)line.at(i) - 64;
-                    break;
-                case 11:
-                    atom.id = static_cast<int>(atoi(value.c_str()));
-                    break;
-                case 17:
-                    atomCode = value;
-                    break;
-                case 79:
-                    atomName = value;
-                    break;
-                case 26:
-                    atom.residue = static_cast<int>(atoi(value.c_str()));
-                    break;
-                case 37:
-                    atom.position[0] = static_cast<float>(atof(value.c_str()));
-                    break;
-                case 45:
-                    atom.position[1] = static_cast<float>(atof(value.c_str()));
-                    break;
-                case 53:
-                    atom.position[2] = static_cast<float>(atof(value.c_str()));
-                    break;
-                default:
-                    if (line.at(i) != ' ')
-                        value += line.at(i);
-                    break;
-                }
-                i++;
-            }
+            return true;
+        }
+        return false;
+    }
 
-            // Material
-            atom.materialId = 0;
-            i = 0;
-            bool found = false;
-            while (!found && i < colorMapSize)
-            {
-                if (atomName == colorMap[i].symbol)
-                {
-                    found = true;
-                    switch (properties.color_scheme)
-                    {
-                    case ProteinLoaderColorScheme::ProteinChains:
-                        atom.materialId = abs(atom.chainId);
-                        break;
-                    case ProteinLoaderColorScheme::ProteinResidues:
-                        atom.materialId = atom.residue;
-                        break;
-                    default:
-                        atom.materialId = static_cast<int>(i);
-                        break;
-                    }
-                }
-                ++i;
-            }
+    static std::vector<Atom> _allocateBuffer(std::string_view content)
+    {
+        auto lines = brayns::StringCounter::countLines(content);
+        auto buffer = std::vector<Atom>();
+        buffer.reserve(lines);
+        return buffer;
+    }
 
-            // Radius
-            atom.radius = DEFAULT_RADIUS;
-            i = 0;
-            found = false;
-            while (!found && i < colorMapSize)
-            {
-                if (atomName == atomic_radii[i].Symbol)
-                {
-                    atom.radius = atomic_radii[i].radius;
-                    found = true;
-                }
-                ++i;
-            }
+    static Atom _processLine(std::string_view line)
+    {
+        auto tokens = brayns::StringSplitter::splitTokens(line);
+        assert(tokens.size() >= 12);
```

Comment:
It could be done without allocation using StringCount::countTokens and StringExtractor::extractToken but it is fine like that also.

But I wouldn't use assert here and throw an exception as we want to check the input files even in release mode...

Otherwise the index access will crash...

Reply:
Addressed